File: javascript/src/Point.js
/**
 * @class
 * A point in a cartesian plane
 * @constructor
 * @param {int} x
 * @param {int} y
 */
function Point(x, y) {
    this.x = x;
    this.y = y;
};

/**
 * Test equality of two points
 * @return boolean
 */
Point.prototype.equals = function(other) {
    return ((this.x == other.x) && (this.y == other.y));
};

/**
 * The string representation of this point.
 * @return string
 */
Point.prototype.toString = function() {
    return "(" + this.x.toString() + ", " + this.y.toString() + ")";
    
};

/**
 * The XML representation of this point
 * @return string
 */
Point.prototype.toXml = function() {
    return "<point x=\"" + this.x + "\" y=\"" + this.y + "\" />";
};

/**
 * length of a vector (the distance of the point from the origin)
 * @return int
 */
Point.prototype.len = function() {
    return Math.round(Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2)));
};

/**
 * The sum of two vectors 
 * @return Point
 */
Point.prototype.add = function(other) {
    return new Point(this.x + other.x, this.y + other.y);
};

/**
 * The difference between two vectors
 * @param {Point} other
 * @return Point
 */
Point.prototype.sub = function(other) {
    return new Point(this.x - other.x, this.y - other.y);
};

/**
 * The dot product of a vector and an integer
 * @param {int} i
 * @return Point
 */
Point.prototype.mul = function(i) {
    return new Point(this.x * i, this.y * i);
};

/**
 * A vector which is a fraction of the original
 * @param {int} i
 * @return Point
 */
Point.prototype.div = function(i) {
    return new Point(Math.round(this.x / i), Math.round(this.y / i));
};

File: python/src/hexmap/point.py
"""
A cartesian point
"""

import math
import lxml.etree as etree

class Point(object):

    def __init__(self, x, y):
        self._x = x
        self._y = y

    @property
    def x(self): return self._x

    @property
    def y(self): return self._y

    def __str__(self):
        return "(%d,%d)" % (self.x, self.y)

    def __len__(self):
        return round(math.sqrt(math.pow(self.x, 2) + math.pow(self.y, 2)))

    def __add__(self, other):
        assert(isinstance(other, Point))
        return Point(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        assert(isinstance(other, Point))
        return Point(self.x - other.x, self.y - other.y)

    def __mul__(self, other):
        assert(isinstance(other, int))
        return Point(self.x * other, self.y * other)

    def __div__(self, other):
        assert(isinstance(other, int))
        assert(other != 0)
        return Point(int(self.x / other), int(self.y / other))
        
    @property
    def element(self):
        e = etree.Element("point")
        e.set("x", self.x)
        e.set("y", self.y)

    @property
    def xml(self): 
        return etree.tostring(self.element, pretty_print=True)

    @classmethod
    def fromelement(cls, element):
        x = int(element.get("x"))
        y = int(element.get("y"))
        return cls(x, y)

    def fromxml(cls, xmlstring):
        element = etree.fromstring(xmlstring)
        return cls.fromelement(element)
